sql: fix sequence reporting of smallint type

cty123 opened the pull request:
Previously the type sequence descriptor was hardcoded to be 'bigint' even when the type is created as INT2. This causes confusion in the 'information_schema' table as reported on cockroach#108847.

In this PR 2 fields are updated:

1. `data_type` field will report `INT2` if it is created as `smallint`(int2), `INtT4`, or `INT8` if it is created as `bigint`(int8).

2. `numeric_precision` for `smallint` will be reported as 16, as smallint only has 16 bits. https://www.postgresql.org/docs/current/infoschema-sequences.html 

fixes https://github.com/cockroachdb/cockroach/issues/108847

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/108878)
<!-- Reviewable:end -->

rafiss reviewed (changes requested):
thanks for your contribution!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108878#-:-NcOnG2pC2vcg5Pm6NPH:b3akoa5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cty123)

___
*[`pkg/sql/information_schema.go` line 1252 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108878#-NcOlO8eEyuBVzS7TFNS:-NcOlO8eEyuBVzS7TFNT:b-a7nllu) ([raw file](https://github.com/cockroachdb/cockroach/blob/792b9a603c9d033b24c5894bdf9bae31414c65d2/pkg/sql/information_schema.go#L1252)):*
> ```Go
> 	case "INT2":
> 		return "smallint"
> 	default:
> ```

this should also have a case for INT4
___
*[`pkg/sql/information_schema.go` line 1260 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108878#-NcOlZJz-hj3uGkfAZzd:-NcOlZJz-hj3uGkfAZze:b-a7nllu) ([raw file](https://github.com/cockroachdb/cockroach/blob/792b9a603c9d033b24c5894bdf9bae31414c65d2/pkg/sql/information_schema.go#L1260)):*
> ```Go
> 	switch integerType {
> 	case "INT2":
> 		return 16
> ```

this should also have a case for INT4
___
*[`pkg/sql/information_schema.go` line 1276 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108878#-NcOli8SAIxcRmildodc:-NcOli8SAIxcRmildodd:b-9zjntr) ([raw file](https://github.com/cockroachdb/cockroach/blob/792b9a603c9d033b24c5894bdf9bae31414c65d2/pkg/sql/information_schema.go#L1276)):*
> ```Go
> 				if !table.IsSequence() {
> 					return nil
> 				}
> ```

i would prefer if you did the calculation inline, rather than as a separate function. this is only used in one place, and that function is not that relevant for anything else.
```
typ := "bigint"
precision := 64
switch table.GetSequenceOpts().AsIntegerType {
case "INT2":
  typ = "smallint"
  precision = 16
case "INT4":
  type = "integer"
  precision = 32
}
```

though this is making be realize that it may be better to use `int2/int4/int8` instead of `smallint/integer/bigint`
___
*[`pkg/sql/logictest/testdata/logic_test/information_schema` line 4646 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108878#-NcOlQazD8HUWiv9uBZp:-NcOlQazD8HUWiv9uBZq:b-v4pbnw) ([raw file](https://github.com/cockroachdb/cockroach/blob/792b9a603c9d033b24c5894bdf9bae31414c65d2/pkg/sql/logictest/testdata/logic_test/information_schema#L4646)):*
> ```
> statement ok
> CREATE SEQUENCE test_seq_3 AS smallint
> ```

this should also have tests for:
```
CREATE SEQUENCE s1 AS INT2

CREATE SEQUENCE s2 AS INT4

CREATE SEQUENCE s3 AS INT8
```
___
*[`pkg/sql/logictest/testdata/logic_test/information_schema` line 4657 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108878#-NcOmTEK2uqTHB0yORuz:-NcOmTEK2uqTHB0yORv-:bla5ygg) ([raw file](https://github.com/cockroachdb/cockroach/blob/792b9a603c9d033b24c5894bdf9bae31414c65d2/pkg/sql/logictest/testdata/logic_test/information_schema#L4657)):*
> ```
> ----
> sequence_catalog  sequence_schema  sequence_name                data_type  numeric_precision  numeric_precision_radix  numeric_scale  start_value  minimum_value  maximum_value        increment  cycle_option
> test              public           generated_as_identity_b_seq  bigint     64                 2                        0              1            1              9223372036854775807  1          NO
> ```

i know postgres uses "bigint" here, but i wonder if it would be more clear if we used "int2", "int4" and "int8" here


<!-- Sent from Reviewable.io -->

cty123 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108878)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/information_schema.go` line 1276 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108878#-NcOli8SAIxcRmildodc:-Ncxd4or9ScYG2Tqn6XL:b-58hw58) ([raw file](https://github.com/cockroachdb/cockroach/blob/792b9a603c9d033b24c5894bdf9bae31414c65d2/pkg/sql/information_schema.go#L1276)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i would prefer if you did the calculation inline, rather than as a separate function. this is only used in one place, and that function is not that relevant for anything else.
```
typ := "bigint"
precision := 64
switch table.GetSequenceOpts().AsIntegerType {
case "INT2":
  typ = "smallint"
  precision = 16
case "INT4":
  type = "integer"
  precision = 32
}
```

though this is making be realize that it may be better to use `int2/int4/int8` instead of `smallint/integer/bigint`
</blockquote></details>

Based on cockroach public documentation, integer is equivalent to int8, not int4, unless this is changed(or is there a plan to change?) https://www.cockroachlabs.com/docs/stable/int. I am in favor of changing this to `int2/int4/int8` instead of `smallint/integer/bigint`, but this would make it incompatible with Postgresql standard https://www.postgresql.org/docs/current/sql-createsequence.html 


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108878)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cty123)

___
*[`pkg/sql/information_schema.go` line 1276 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108878#-NcOli8SAIxcRmildodc:-Ncxe2aNEvKZvhHLQOIg:b-gynq8p) ([raw file](https://github.com/cockroachdb/cockroach/blob/792b9a603c9d033b24c5894bdf9bae31414c65d2/pkg/sql/information_schema.go#L1276)):*
> integer is equivalent to int8, not int4

Further down on that docs page, it mentions the `default_int_size` setting. That can change the meaning of `integer`.
```
root@localhost:26257/defaultdb> set default_int_size = 4;
SET

root@localhost:26257/defaultdb> create table t(a integer);
CREATE TABLE

root@localhost:26257/defaultdb> show create table t;
  table_name |                      create_statement
-------------+--------------------------------------------------------------
  t          | CREATE TABLE public.t (
             |     a INT4 NULL,
             |     rowid INT8 NOT VISIBLE NOT NULL DEFAULT unique_rowid(),
             |     CONSTRAINT t_pkey PRIMARY KEY (rowid ASC)
             | )
(1 row)
```

In this case, I think it's OK to deviate from postgres.


<!-- Sent from Reviewable.io -->

cty123 committed:
sql: fix sequence reporting of smallint type

Previously the type sequence descriptor was hardcoded to be 'bigint' even when the type is created as INT2. This causes confusion in the 'information_schema' table as reported on cockroach#108847. To make typing description explicit, we switch data type reporting from 'smallint/integer/bigint' used by PostgreSQL to 'INT2/INT4/INT8'.

Fixes cockroachdb#108847

Epic: none

Release note: None

rafiss commented:
thank you for your contribution!

bors r+
